cirq-core/cirq/ops/controlled_gate.py

```diff
@@ -159,6 +159,16 @@ def _decompose_with_context_(
         self, qubits: Tuple['cirq.Qid', ...], context: Optional['cirq.DecompositionContext'] = None
     ) -> Union[None, NotImplementedType, 'cirq.OP_TREE']:
         control_qubits = list(qubits[: self.num_controls()])
+        controlled_sub_gate = self.sub_gate.controlled(
+            self.num_controls(), self.control_values, self.control_qid_shape
+        )
+        # Prefer the subgate controlled version if available
+        if self != controlled_sub_gate:
+            # Prevent 2-cycle from appearing in recursive decomposition
+            if not isinstance(controlled_sub_gate, ControlledGate) or not isinstance(
```

Comment:
Is this condition possible? The constructor for ControlledGate absorbs the control layers if the subgate is another ControlledGate. Do any of the unit tests fail if this check is removed?

Reply:
Yes, I tried first without and the unit tests were failing.

I discovered that a 2-cycle forms in the decomposition for something like CCZ and control_qid_shape = [3].
The following happens:
* when the sub gate is CZ, `CZPowGate.controlled()` sees that `control_qid_shape[-1] != 2` and returns a `ControlledGate`. Then `self != controlled_sub_gate` is false. Next, the special case in `ControlledGate._decompose_with_context_()` for `CZPowGate` creates a `Z` gate and adds +1 to the control qubits;
* when the sub gate is Z, the `ZPowGate.controlled()` method returns a `CZPowGate` and subtracts -1 from the control qubits. Then `self != controlled_sub_gate` is true.

This cycles infinitely, adding and removing a control qubit.